Use 'cmake --install' where available

KazNX opened the pull request:
Make use of CMake '--install' option introduced in 3.15. Only performs installation, whereas '--build <dir> --target install' would first (re)build. Using '--install' is a significantly faster installation process when using Visual Studio as various overheads are avoided.

dirk-thomas commented:
Builds on top of #56. Waiting for that one to be merged first.

dirk-thomas commented:
With #56 merged can you please rebase this patch.

KazNX commented:
I've resolve the merge conflicts, however, I merged rather than rebasing.

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +209,53 @@ def get_visual_studio_version():
+                    int(ver_match.group(3)), ver_match.group(4))
+
+    # Failed to parse the version number.
+    return (0, 0, 0, 'failed')
```
Instead of implementing a custom function how about using `pkg_resources.parse_version` instead?

KazNX replied:
Good to know.

KazNX replied:
Right. In which case the following test will fail because the major part of the version is less than 3. At any rate, this is moot with the change to using pkg_resources for version parsing.

dirk-thomas commented on colcon_cmake/task/cmake/build.py:
```diff
@@ -313,16 +314,30 @@ def _get_make_arguments(self):
-        cmd = [
-            CMAKE_EXECUTABLE, '--build', args.build_base,
-            '--target', 'install']
+        cmake_ver = await get_cmake_version()
```
Currently the return value is not being checked for a failure.

KazNX replied:
Do you mean failure to parse - i.e, checking for a version number result which indicates failure -  or for failure of the await directive?

dirk-thomas replied:
Your custom function might return `(0, 0, 0, 'failed')`.

KazNX replied:
The return values have changed. This is no longer relevant.

dirk-thomas replied:
While the way the version is parsed has changed the fact that the parsing can fail for an invalid string hasn't. And that case should be explicitly handled.

dirk-thomas commented on colcon_cmake/task/cmake/build.py:
```diff
@@ -313,16 +314,30 @@ def _get_make_arguments(self):
+                # Arguably, _get_make_arguments() should be be configured to
+                # return arguments based on the current generator, not just for
+                # make (and ninja by coincidence). The '--' extension may be
+                # better returned from _get_make_arguments()
```
Please remove those comments which have been removed from the other PR.

KazNX replied:
Will do. Just catching up on various comments.

dirk-thomas commented on colcon_cmake/task/cmake/build.py:
```diff
@@ -313,16 +315,26 @@ def _get_make_arguments(self):
+            # This only installs, whereas --build <dir> --target install will
+            # build (again) then install.
+            cmd += ['--install', args.build_base]
+        else:
```
Why is choosing the `else` block in case the version couldn't be retrieved the right choice?

KazNX replied:
`--install` is a newer feature requiring 3.15 whereas the else statement uses `--build --target` which has had long term support. We can only guarantee support for `--install` if we can validate that the version number is `> 3.15`.

dirk-thomas replied:
I understand that but are you expecting the get version function to fail for any CMake version? If not, please check for the potential returned error and not just silently use the else block.

KazNX replied:
Ok. I'll add an error message in the case wheret eh CMake version cannot be determined.

KazNX commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +210,29 @@ def get_visual_studio_version():
+    ver_line = lines[0] if lines and len(lines) else None
+    if ver_line and len(ver_line):
+        # Extract just the version part of the string.
+        ver_re_str = r'^(?:cmake\s+version)?\s*(.*)$'
```
Excellent question. I have no idea. If I don't remove the "cmake version" prefix, then the version number comes out as a LegacyVersion object, which will never pass the later checks for 3.15+. This means I have to do some parsing of the line here.

What do you think of using the following regular expression then: `r'^(?:.*)(\d+\.\d+\..*)'`. This reduces assumptions about the string. The requirements is that the pattern `#.#.<anything>` is present where `#` are numeric sequences.

dirk-thomas replied:
What is the problem with comparing to a `LegacyVersion`?

KazNX replied:
Running `cmake --verison` returns a string in the form: `cmake version 3.15.0`, optionally ending in `-rc1` for release candidates (at least in English). Simply reading the line as is as giving that to `pkg_resources.parse_version()` wil always generate a `LegacyVersion` because the string isn't just a version number. As I see it, `LegacyVersion` is not comparable; that is it is equivalent to version 0. I need to drop `cmake version ` in order to generate an informative version number object. The expression I propose has fewer assumptions about the cmake version string, assuming only that it ends in a `#.#.<text>` form. This allows any prefix to be on the version number line, regardless of language.

KazNX replied:
Note: resolving this issues will also resolve the other oustanding issues.

KazNX commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +210,29 @@ def get_visual_studio_version():
+        if ver_match:
+            return parse_version(ver_match.group(1))
+    # Failed: return invalid version
+    return parse_version('0.0.0')
```
Will fix.

KazNX replied:
Fix is m

dirk-thomas commented:
Thanks for iterating on this. I have one more comment though: atm each package will determine the CMake version separately. It would be good if the function persists the determined version in a global variable and return that directly when being involved repeatedly.

dirk-thomas commented:
@KazNX Are you planning on addressing the last comment (determining the CMake version only once)?

KazNX commented:
@dirk-thomas Sorry, I've been busy on deliverables. Can you propose the best way to determine the CMake globally?

dirk-thomas commented:
> Can you propose the best way to determine the CMake globally?

@KazNX Please see my previous comment for that:

> It would be good if the function persists the determined version in a global variable and return that directly when being involved repeatedly.

See https://github.com/colcon/colcon-ros/blob/ba7472246ec4235fcae0ce71b055aff07bfec4c4/colcon_ros/task/cmake/__init__.py#L33-L59 for an example.

dirk-thomas commented:
@KazNX Friendly ping.

KazNX commented:
I have a caching solution in place now.

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +210,55 @@ def get_visual_studio_version():
+
+
+# Global variable for the cached CMake version number.
+_cached_cmake_version = _parse_cmake_version()
```
Currently this determines the CMake version at import time of this module. Please change it to be lazily computed on the first invocation of `get_cmake_version` (as in the previously referenced example).

KazNX replied:
Now lazy.

BTW, should I be clicking `Resolve conversation` or should you?

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +210,72 @@ def get_visual_studio_version():
+    try:
+        output = subprocess.check_output([CMAKE_EXECUTABLE, '--version'])
+        lines = output.decode().splitlines()
+        ver_line = lines[0] if lines and len(lines) else None
```
The condition `len(lines)` is redundant with `lines`.

The logic can also be reduced by removing this line, changing the condition in the next line to just `if lines:`, and then below use `lines[0]` instead of `ver_line`.

KazNX replied:
Fixed

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +210,72 @@ def get_visual_studio_version():
+
+    The result is None when a version number is not available or cannot be
+    parsed. This may occur if CMake is not present, is a very old version or is
+    a newer version where the version number string output has changed.
```
For what never version is that the case? Then the `--install` option wouldn't be used.

KazNX replied:
I'm identifying the behaviour should the cmake version string format change in future. More of a maintenance note.

KazNX replied:
I've modified the comment slightly.

KazNX commented:
All issues should be addressed now.

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +211,64 @@ def get_visual_studio_version():
+        lines = output.decode().splitlines()
+        if lines:
+            # Extract just the version part of the string.
+            ver_re_str = r'^(?:.*)(\d+\.\d+\..*)'
```
I am not sure I understand why this regex was chosen. Can you please elaborate. It would be good to add a unit test to ensure this function does what it is supposed to do.

KazNX replied:
It does look odd starting with a non capturing group. I think this is also a throw back to when I had "cmake version" as part of the expression. Really I'm looking for `#.#.anything` where `#` is a number. CMake version numbers will often end with `.1rc2` or the like, so the patch number can't be read purely numerically.

dirk-thomas replied:
> CMake version numbers will often end with .1rc2 or the like, so the patch number can't be read purely numerically.

I still don't understand why the regex would end in `\d+\..*`? In your example there is no dot between `.1` and `rc2`.

dirk-thomas replied:
@KazNX Friendly ping.

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +211,64 @@ def get_visual_studio_version():
+        if lines:
+            # Extract just the version part of the string.
+            ver_re_str = r'^(?:.*)(\d+\.\d+\..*)'
+            ver_match = re.match(ver_re_str, lines[0], re.I)
```
Why use `re.I` here? The pattern doesn't seem to contain any alphabetical characters.

dirk-thomas replied:
@KazNX Are planning to update the patch accordingly?

dirk-thomas commented:
Also trying this patch with a CMake version new enough to trigger `--install` the invocation fails for me on Linux since it passes the `job_args` which seems to be not supported for the install option.

dirk-thomas commented:
@KazNX Friendly ping.

KazNX commented:
@dirk-thomas I've been waiting for responses on my comments. I'll have a look at the linux issue and streamlining the regex. Is there a simple example I can follow for the unit test?

dirk-thomas commented:
> I've been waiting for responses on my comments.

Can you please clarify what comments you are referring to?

As far as I can see I was the last person commenting on all pending conversation threads:
* regex pattern: https://github.com/colcon/colcon-cmake/pull/57#discussion_r364570229
* regex flag: https://github.com/colcon/colcon-cmake/pull/57#discussion_r367072807
* failing on Linux (and likely macOS): https://github.com/colcon/colcon-cmake/pull/57#issuecomment-574830801

> Is there a simple example I can follow for the unit test?

I am not sure what you are referring to? Regarding the failing build you should just try to build any CMake package using your patch.

KazNX commented:
>> I've been waiting for responses on my comments.

> Can you please clarify what comments you are referring to?

Looks like this is an unfamiliarity with GitHub on my part. I need to hit submit.


- regex parttern simplified
- regex flag removed
- Linux job_args removed

>> Is there a simple example I can follow for the unit test?

> I am not sure what you are referring to? Regarding the failing build you should just try to build any CMake package using your patch.

I was after an example to follow as I'm unsure how you are invoking unit tests. I added `test/test_parse_cmake_version.py` and it is failing the build because it can't import `colcon_cmake`. I'm not familiar with the cause.

dirk-thomas commented on colcon_cmake/task/cmake/build.py:
```diff
@@ -333,10 +333,8 @@ def _get_make_arguments(self):
         if multi_configuration_generator:
             cmd += ['--config', self._get_configuration(args)]
         else:
-            # Add make/ninja arguments for the number of job threads to use.
```
Looking at the diff of this PR that doesn't seem to be the case.

dirk-thomas commented:
> `^(?:.*)` is questionable now. It's a non-capturing group. Better to replace with `.*`.

`.*(\d+\.\d+)` isn't correct since for an input string like `foo 12.3` the captured group will be only `2.3` because the leading wildcard is greedy. How about checking for the whitespace before the digit?

> `(\d+\.\d+\..*)` matches `(#.#.<anything>)` to match `major.minor.patch+comment`

The trailing `.*` matches anything. Could there be a "+comment" which you can't parse as a version? Why not limit the capture group to end after the digits?

> I'm unsure how you are invoking unit tests.

Just invoking `pytest` - the coverage flags are optional, see https://github.com/colcon/colcon-cmake/blob/c3e809cd57e38d10c15fa80df7fe0876252956a1/.travis.yml#L14

> I added `test/test_parse_cmake_version.py` and it is failing the build because it can't import `colcon_cmake`. I'm not familiar with the cause.

Locally the linters fail with the following:

```
test/test_copyright_license.py:10
In some files no copyright / license line was found
```

as well as 

```
1     W292 no newline at end of file
1     E302 expected 2 blank lines, found 1
1     E305 expected 2 blank lines after class or function definition, found 1
1     E501 line too long (80 > 79 characters)
2     F401 'pkg_resources.parse_version' imported but unused
1     Q000 Remove bad quotes
```

---

Regarding the comments: lets first focus on getting the functional logic of the patch right and working as well as the tests turn over. Then we can revisit the comments and other nitpicks.

KazNX commented:
I've updated the regular expression to be less greedy with leading characters. I'm also ignoring any non-numeric suffix. It turns out CMake can have many suffixes including '-dirty' on working versions. I don't see it being worthwhile to distinguish these. I've tried a number of expected and some unexpected strings in `test_parse_cmake_version.py` including checking the greedy case you've highlighted.

I've fixed the linter issues. I didn't notice these because it's the following error which has my attention:

> ImportError while importing test module '/home/travis/build/colcon/colcon-cmake/test/test_parse_cmake_version.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> test/test_parse_cmake_version.py:4: in <module>
>     from colcon_cmake.task.cmake import parse_cmake_version_string
> E   ImportError: No module named 'colcon_cmake'

Clearly it's failing to import the module I'm trying to test. This is where I'm unsure of why though. I can run the test from the command line. By contrast, colcon-core tests do something similar, importing modules from within colcon-core.

dirk-thomas commented:
> Clearly it's failing to import the module I'm trying to test. This is where I'm unsure of why though. I can run the test from the command line. By contrast, colcon-core tests do something similar, importing modules from within colcon-core.

Please try to create an empty file `test/__init__.py`.

KazNX commented:
That seems to have done the trick. Thanks. That should address all your oustanding issues.

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -215,24 +215,22 @@ def get_visual_studio_version():
 def get_cmake_version():
     """
-    Get the cached CMake version number if available and successfully parsed.
+    Get the CMake version.
```
I agree. Currently the additional information if in the following paragraph. Please feel free to suggest a better first sentence (which passes the linter in terms of line length).

KazNX replied:
If I'm right with the line limit being 80 characters, then the orignal comment fit. Otherwise, how about:

> Query the CMake --version results if available.

Otherwise, I'm not going to deliberate over this one. I'd like to complete the PR.

dirk-thomas replied:
I wouldn't use "query". Also "results" doesn't sound right to me. And "if available" is fuzzy too.
How about:

> Get the version as reported by `CMAKE_EXECUTABLE --version` if parsable.

The redundant detail can then be removed from `:returns:`:

> The CMake version, or None when the version number could not be determined

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -262,23 +260,9 @@ def get_cmake_version():
-    cache the CMake version number.
-
-    External API users should use get_cmake_version()
+    Parse the CMake version printed by `CMAKE_EXECUTABLE --version`.
```
Imo some of the sentences where not helpful, e.g.:

> This function blocks on the execution of cmake and should only be used to cache the CMake version number.

Since the function is not `async` it clearly is blocking. I don't see why that should be part of the docblock.

---

> Parsing is performed by the pkg_resources package and the returned object is a Version object from that package.

The return type clearly documents how the version is represented.

---

> The version number is parse from the output of 'cmake --version' and is expected in the form 'cmake version #.#.#' where each '#' character represents part of the version number.
> Additional text may follow, however only the first line is parsed.

The part - that the output of `cmake --version` is parsed - is imo the important part for the reader. The other text isn't relevant to the user of the API - at most it could be a comment documenting how the logic works if a developer wants to read / understand the implementation of the function.

KazNX replied:
Ok. Again, I'm not going to hold the PR up on these details.

dirk-thomas commented on colcon_cmake/task/cmake/build.py:
```diff
@@ -325,13 +325,10 @@ def _get_make_arguments(self):
-                logger.error(
-                    'Unable to determin CMake version. Fallback to building '
-                    'the install target in place of cmake --install.')
+                logger.warning(
```
Can you please clarify why `not cmake_ver` would be `True` and this warning be triggered if a project has no install target?

KazNX replied:
I will be `False` on failure to parse rather than `None`. This is so the upstream doce can distuguish between no parsing attempt made (`None`) and failed to parse (`False`).

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +211,81 @@ def get_visual_studio_version():
+
+    if _cached_cmake_version is False:
+        # in case the CMake version could not be determined
+        return None
```
The logic around the cache value is pretty convoluted coming from the fact that the function uses a different "negative" value (`None`) than the cached value (`False`).

I aligned those in 6390dd148c23f910309afe802c76cbdc2dddff3d which gets rid of the majority of the logic in this function.

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +211,71 @@ def get_visual_studio_version():
+    """
+    Parse the given CMake version string.
+
+    This is for internal use and testing.
```
Since the function is meant to be internal I removed this line from the docblock and instead prefixed the function name with an underscore: 631e008b53811e6031dc42cb30d3d1b6db27cf15.

dirk-thomas commented on test/test_parse_cmake_version.py:
```diff
@@ -0,0 +1,44 @@
+        else:
+            assert parsed_version._version.release[0] == expected_ver[0]
+            assert parsed_version._version.release[1] == expected_ver[1]
+            assert parsed_version._version.release[2] == expected_ver[2]
```
Simplified the test logic in 83bd8e76dc6e4dabca5fecbdb48f7f3933658a14.

dirk-thomas commented on colcon_cmake/task/cmake/__init__.py:
```diff
@@ -208,3 +211,69 @@ def get_visual_studio_version():
+    :rtype pkg_resources.extern.packaging.version.Version
+    """
+    # Extract just the version part of the string.
+    ver_re_str = r'^(?:.*\s)?(\d+\.\d+\.\d+).*'
```
Using `\s` for matching whitespace: see 83bd8e76dc6e4dabca5fecbdb48f7f3933658a14.

dirk-thomas commented on test/test_parse_cmake_version.py:
```diff
@@ -0,0 +1,40 @@
+# Copyright 2016-2018 Dirk Thomas
```
Please consider use your name for the copyright of this file since you wrote it. Also the year should 2020.

KazNX replied:
Fixed

dirk-thomas commented:
Some ROS 2 CI builds to double check:
* Linux [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=9172)](https://ci.ros2.org/job/ci_linux/9172/)
* macOS [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_osx&build=7485)](https://ci.ros2.org/job/ci_osx/7485/)
* Windows [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9088)](https://ci.ros2.org/job/ci_windows/9088/)

KazNX commented:
> Some ROS 2 CI builds to double check:

They all look good, but are they running with this change or does something need to be done in order to test this PR with those builds?

dirk-thomas commented:
> are they running with this change

Yes, I have pushed the changes from this PR to a branch within this repo (https://github.com/colcon/colcon-cmake/tree/enhancement/use-cmake-install) and the jobs are using that branch (see CI parameter `colcon_branch`).

This is good to be merged. Thank you for this nice improvement. I am sorry for the multiple rounds of iteration and appreciate that you kept iterating on it. :bowing_man:
